core/store/src/flat_state.rs

```diff
@@ -9,34 +9,134 @@
 //! reference, one to get value itself.
 // TODO (#7327): consider inlining small values, so we could use only one db access.
 
+use borsh::{BorshDeserialize, BorshSerialize};
+
 #[cfg(feature = "protocol_feature_flat_state")]
 mod imp {
+    use borsh::BorshSerialize;
+
+    use near_primitives::block_header::BlockHeader;
     use near_primitives::errors::StorageError;
     use near_primitives::hash::CryptoHash;
+    use near_primitives::shard_layout::ShardUId;
     use near_primitives::state::ValueRef;
 
-    use crate::Store;
+    use crate::flat_state::KeyForFlatStateDelta;
+    use crate::{DBCol, FlatStateDelta, Store, StoreUpdate};
 
     /// Struct for getting value references from the flat storage.
     ///
     /// Used to speed up `get` and `get_ref` trie methods.  It should store all
     /// trie keys for state on top of chain head, except delayed receipt keys,
     /// because they are the same for each shard and they are requested only
     /// once during applying chunk.
-    // TODO (#7327): implement flat state deltas to support forks.
-    // TODO (#7327): store on top of final head (or earlier) so updates will
-    // only go forward.
+    // TODO (#7327): lock flat state when `get_ref` is called or tail is being updated. Otherwise, `apply_chunks` and
+    // `postprocess_block` parallel execution may corrupt the state.
     #[derive(Clone)]
     pub struct FlatState {
+        /// Used to access flat state stored at the head of flat storage.
+        /// It should store all trie keys and values/value refs for the state on top of
+        /// flat_storage_state.head, except for delayed receipt keys.
         store: Store,
+        /// Id of the shard which state is accessed by this object.
+        shard_uid: ShardUId,
```

Comment:
I thought about it more and using `shard_id` here is enough, we don't need `shard_uid`. 

When there are shard migration and we need to apply a block to two shard layouts, we actually still only apply transactions only once. Let's say that shard with shard uid (0, 0) is split to 2 shards (1, 0) and (1, 1), apply transactions will only called on shard (0, 0), then we split the state changes happened in this block and apply that to the tries of shards (1, 0) and (1, 1). Since transactions only need to be applied to shard uid (0, 0), we only need to maintain flatstorage for this shard. That means, we only need to maintain flat storage for the shard layout the chain is currently using. 

To fully support sharding migration, we will need to implement some logic at epoch boundary to split/merge the FlatStorageDeltas from the old shard layout to the new shard layout, but we don't need shard_uid for that. It is also important that the key for `FlatState` does not include `shard_id` like in `Trie` (I don't think it's the case now but just want to mention it), otherwise we need to reconstruct the whole `FlatState` column when resharding happens.